File: May-LeetCoding-Challange-2020/Week4/986-interval-list-intersections.js
/**
 * @param {number[][]} A
 * @param {number[][]} B
 * @return {number[][]}
 * Runtime: 92 ms, faster than 40.48% of JavaScript online submissions for Interval List Intersections.
 * Memory Usage: 43.2 MB, less than 100.00% of JavaScript online submissions for Interval List Intersections.
 * TC: O(A.length + B.length) which is linear
 * SC: O(1) if we don't include result
 */
var intervalIntersection = function(A, B) {
    if (!A || !B) {
        return [];
    }

    let a = 0, b = 0;
    const result = [];
    while (a < A.length && b < B.length) {
        if (doesOverlap(A[a], B[b])) {
            result.push([Math.max(A[a][0], B[b][0]), Math.min(A[a][1], B[b][1])]);

            if(A[a][1] <= B[b][1]) {
                a++;
            } else {
                b++;
            }
        } else if(A[a][0] > B[b][1]) {
            b++;
        } else {
            a++;
        }
    }

    return result;
};

function doesOverlap (range1, range2) {
    const [start1, end1] = range1;
    const [start2, end2] = range2;

    return end2 >= start1 && end1 >= start2;
}

File: May-LeetCoding-Challange-2020/Week4/986-Interval-List-Intersections.cpp
// Time Complexity: O(n+m) has to go through A and B all the way
// Space Complexity: O(n+m) if A and B are equal
// Runtime: 60 ms, faster than 93.71%
// Memory Usage: 18.8 MB, less than 8.00%

// Best Case == Worst Case
class Solution {
public:
    vector<vector<int>> intervalIntersection(vector<vector<int>>& A, vector<vector<int>>& B) {
        int start, end;
        int i = 0, j = 0;
        vector<vector<int>> result;
        
        while(i < A.size() && j < B.size()){
            start = max(A[i][0], B[j][0]);
            end = min(A[i][1], B[j][1]);
            
            if(start <= end)
                result.push_back({start, end});
            
            if(B[j][1] > A[i][1]){
                i++;
            }else{
                j++;
            }
        }
        
        return result;
    }
};